src/rgw/rgw_common.h

```diff
@@ -1633,6 +1635,8 @@ struct req_state : DoutPrefixProvider {
   string bucket_instance_id;
   int bucket_instance_shard_id{-1};
   string redirect_zone_endpoint;
+  std::stack<Span> stack_span;
```

Comment:
@cbodley did not get this

Reply:
in normal code, you're passing Spans around for use in `child_span()`, ie:
```c++
int main() {
  auto span = new_span("main");
  func1(span);
}
void func1(const Span& parent_span) {
  auto span = child_span("func1", parent_span);
  func2(span);
}
void func2(const Span& parent_span) {
  auto span = child_span("func2", parent_span);
  func3(span);
}
void func3(const Span& parent_span) {
  auto span = child_span("func3", parent_span);
  ...
}
```
by the time func3() runs, there are 4 Spans on the call stack, and every function knows exactly which one is their parent

but with the `req_state_span`, you're relying on a `std::stack` to find the next parent instead. `std::stack` will be allocating memory, so is more expensive than using the call stack; so i'd suggest using the call stack everywhere and drop this `req_state_span` stuff

Reply:
the main advantage I m getting with stack_span method is that I dont have to pass around span as a parameter, which makes it easier as I dont have to do change function signature (with their virtual function), also I there is no risk of leaving out some function if I accidentally dont supply it with a parent_span,and also reduced code length.

I will do it though it will take time as I have to change many things I will update soon :)

Reply:
@cbodley makig `Span* parent_span` inside req_state which will always hold the current parent_span for a particular function, tested it with createBucket and it is working fine, this approach wont take any allocation and also not requiring to overload the functions with optional span parameter

Reply:
@cbodley sorry for the counter argument, 

I tried to overcome this approach by ading a new data member in `req_state`, `Span parent_span` it was working fine but in request like RGWGetObj and PutObj it was failing the approach was very dangerous as it invloves moving raw pointers of unique_ptr (`Span`), deferencing those and all those dangerous pointers stuff, so I dropped it.

I started with second approach today that is passing spans around as function parameter, but this will create some very heavy changes as I m needed to change the function signatures and those functions are not single lone functions, majority of them are either overloaded virtual function, or they themseleves are the virtual functions that are overloaded somewhere else, and functions in file `rgw_op.h` are  used in many other files so I need to change all of them even the ones which doesnt need tracing for examples in files like `rgw_rest_pubsub`, `rgw_sync_module_pubsub_rest.cc`. This(according to me) is adding to much code which doesnt serve any purpose.

Now coming to memory part, the stack is only used where function requires tracing and has access to req_state, now function likes `op->pre_exec, op->verify_permission()` all are the child of `root_span`, they are inserted inside the stack but they are soon retrieved back as these functions hardly call any other functions inside them so for these stack size if 1, `op->execute functions calls` functions like send_response_data and all, and send_response_data does call any other function  so then my stack size is 2, so at most(for safe size) I can assume the maximum stack_span size I m getting is 3-4, I dont know if this affects the performance but it is quite low :).
Functions like `init_permission, read_permission, ` calls bucket build and object build policies and they themselcves call one more function `bucket/objectread_policy` so lets take atmost 6 spans of total in `stack_span`.

I m just trying to say the this current impl seems fine( just me), and changing these have good amount of code added to source.